sdk/core/src/context.rs

```diff
@@ -1,11 +1,29 @@
+use chrono::{DateTime, Utc};
+
 /// Pipeline execution context.
 #[derive(Clone, Debug, Default)]
 pub struct Context {
-    _priv: (),
+    timeout: Option<DateTime<Utc>>,
```

Comment:
Is there a general runtime standard / common practice for either HTTP clients we can abstract, or futures we can use instead? Those same timeout mechanisms we use in other languages are also the same mechanisms for manually canceling, like hooking up to a Ctrl+C handler for console apps. How that is done may vary from language/runtime to another.

Reply:
## Cancellation in Rust

Cancellation in Rust is done by dropping the future currently being executed. Timeouts can be performed by racing a client request with a timeout future, and signal-based cancellation can be performed by racing with an abort signal.

I'm omitting the exact imports of these libraries since we'd expect end-users to bring their own. But here's an assortment of examples of how Futures can be cancelled:

__base case: no cancellation__
```rust
let res = surf::get("https://example.com").await?;
println!("Response received: {}", res);
```

__timeout for duration__
_Golang equivalent: Context.WithTimeout_
```rust
let res = surf::get("https://example.com").timeout(Duration::secs(3)).await??;
println!("Response received: {}", res);
```
__timeout until instant__
_Golang equivalent: Context.WithDeadline_
```rust
let instant = Instant::now() + Duration::from_secs(1);
let res = surf::get("https://example.com").timeout_at(instant).await??;
println!("Response received: {}", res);
```

__stop on signal__
_Golang equivalent: Context.WithCancel_
```rust
let src = StopSource::new(); // sends a cancellation signal to all tokens when dropped
let stop_signal = src.token();
let res = surf::get("https://example.com").until(stop_signal).await??;
println!("Response received: {}", res);
```

## Comparing to Golangs Context

The examples above cover most of what Go's `Context` object does. Because cancellation propagates by default for futures, we do not need to pass a context into the client to support cancellation. The only case Go's context provided we're not covering in our examples here is the key-value storage aspect of it. That reason alone seems important enough to keep passing in a context, but unlike Go we do not also need to attach cancellation propagation to it.

## Do we need to do anything else in Rust

Depending on which HTTP client we use (e.g. `hyper`, `reqwest`) we need to ensure that the ongoing HTTP request is cancelled on drop. I don't know either client too well, but in case "cancel on drop" isn't supported in either (which is required to make cancellation correctly propagate), we could author a wrapper [similar to this](https://gist.github.com/yoshuawuyts/77a2f4304318d0e7e16547dc69152909) to fix the cancellation propagation.

Rust's design of futures is very well suited towards cancellation, and removes the need for us to employ some of the methods employed by other languages. All we need to do is ensure that cancellation is correctly propagated at the points where we bring in async runtimes, and we should be good!

Reply:
But what exactly does it mean to drop the future? Seems it would just not call back to the execution context but the current call may proceed. I can't say with assuredness for all languages, but this may not be enough. Consider uploading a large blob to Storage. Canceling should not simply complete the upload in the background, IMO, but actually cancel the executing request i.e. stop uploading bits.

Reply:
I've read this now and I think I understand it better. Having customer code examples would be helpful. So I think a customer would do something like this:
main () {
   let fut = SomeFunc(...)
   fut.until(5 seconds)
}

async fn SomeFunc(...) {
   let fut = OtherFunc(...)
   fut.until(10 seconds)
}

In this case, when does main timeout? I'm guessing 5 seconds (which is good).
In this case when does SomeFunc timeout? I'd like 5 seconds but I think its 10 seconds.
In our ideal world, the lowest time duration should cancel all things "under it".